Touch Bar Simulator/util.swift

```diff
@@ -47,12 +47,117 @@ extension NSWindow {
 	}
 }
 
+extension NSWindow {
+	enum MoveXPositioning {
+		case left, center, right
+	}
+	enum MoveYPositioning {
+		case top, center, bottom
+	}
+
+	func moveTo(x xPositioning: MoveXPositioning, y yPositioning: MoveYPositioning) {
+		let visibleFrame = NSScreen.main!.visibleFrame
+
+		let x: CGFloat, y: CGFloat
+		switch xPositioning {
+		case .left:
+			x = visibleFrame.minX
+		case .center:
+			x = visibleFrame.midX - frame.width / 2
+		case .right:
+			x = visibleFrame.maxX - frame.width
+		}
+		switch yPositioning {
+		case .top:
+			y = visibleFrame.maxY - frame.height
+		case .center:
+			y = visibleFrame.midY - frame.height / 2
+		case .bottom:
+			y = visibleFrame.minY
+		}
+
+		setFrameOrigin(CGPoint(x: x, y: y))
+	}
+}
+
 extension NSView {
 	func addSubviews(_ subviews: NSView...) {
 		subviews.forEach { addSubview($0) }
 	}
 }
 
+final class AssociatedObject<T: Any> {
+	subscript(index: Any) -> T? {
+		get {
+			return objc_getAssociatedObject(index, Unmanaged.passUnretained(self).toOpaque()) as! T?
+		} set {
+			objc_setAssociatedObject(index, Unmanaged.passUnretained(self).toOpaque(), newValue, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
+		}
+	}
+}
+
+@objc
+protocol TargetActionSender: AnyObject {
+	var target: AnyObject? { get set }
+	var action: Selector? { get set }
+}
+
+extension NSControl: TargetActionSender {}
+extension NSMenuItem: TargetActionSender {}
+
+private final class ActionClosureCaller: NSObject {
+	@objc
+	func callClosure(_ sender: TargetActionSender) {
+		onAction?(sender)
+	}
+
+	var onAction: TargetActionSender.ActionClosure?
+}
+
+struct TargetActionSenderAssociatedKeys {
+	fileprivate static let caller = AssociatedObject<ActionClosureCaller>()
+}
+
+extension TargetActionSender {
+	typealias ActionClosure = ((TargetActionSender) -> Void)
+
+	/**
+	Closure version of `.action`
+
+	```
+	let menuItem = NSMenuItem(title: "Unicorn")
+	menuItem.onAction = { sender in
+		print("NSMenuItem action: \(sender)")
+	}
+	```
+	*/
+	var onAction: ActionClosure? {
+		get {
+			return TargetActionSenderAssociatedKeys.caller[self]?.onAction
+		}
+		set {
+			if let caller = TargetActionSenderAssociatedKeys.caller[self] {
+				caller.onAction = newValue
+				action = #selector(ActionClosureCaller.callClosure)
+				target = caller
+			} else {
+				TargetActionSenderAssociatedKeys.caller[self] = ActionClosureCaller()
+				self.onAction = newValue
+			}
+		}
+	}
+}
+
+extension NSApplication {
+	func leftMouseIsDown() -> Bool {
+		return currentEvent?.type == .leftMouseDown
+	}
+
+	func optionKeyIsDown() -> Bool {
```

Comment:
These would be better as getters than functions.

`leftMouseIsDown` => `isLeftMouseDown`